### PR TITLE
Cross-platform CI scripts

### PR DESCRIPTION
**Pull request by aminya**
### Description of the change
- This uses Powershell 7 as a cross-platform shell for CI scripts:
	- PowerShell 7 allows using the same syntax and code across all the operating systems.
	- This simplifies CI scripts. For example, previously CMD was used in Windows, which is an old shell with limited functionality.
	- The old linux_trusty container did not detect pwsh, so we use the Linux image provided by Microsoft, and we use Clang 9 that is provided in it.

- This refactors CI scripts into separate files (templates): 
	- It also allows using parallelization of the tests without copy-pasting the same code. For example: 
		- In macOS, the same templates are reused in the test phase. 
		- In Windows x86, the templates are used for parallelization of tests in [this PR](https://github.com/atom-ide-community/atom/pull/31)

	- This makes it easier to manage CI code. For example, we can quickly compare the publish step for all operating systems without going through big files and trying to find the relevant steps.

Previously, the CI scripts were written back in ~2017. At that time, Azure Templates had not been introduced yet, and because of that using PowerShell was not suitable yet. Now in 2020, we can benefit from these and improve the CI. This is also a requirement for the future Test Parallelization pull request.

### Verification
The CI passes. Tested in atom-ide-community

### Release Notes
N/A

~### Includes~
~This includes #21023.~
~This supersedes #21057.~

~### Thing you should do~
~Because this includes #21023, it requires you to set `SHOULD_SIGN=true` as an environment variable inside Azure.~

### Fork Pull request
https://github.com/atom-ide-community/atom/pull/46

### CODE REVIEW

**Comment by sadick254** [comment]
@aminya Great job 👍 

Since this is a huge refactor I would recommend splitting it into multiple PR's(1 for each template). That way it becomes easy to review and track the impact the changes have on our CI.

Am happy to assist with that if need be.

**Comment by aminya** [comment]
Start by merging https://github.com/atom/atom/pull/21023. The commits are granular, so they can be grouped together and cherry-picked if needed.

**Comment by sadick254** [comment]
@aminya I see this now has conflicts that need to be resolved.

**Comment by aminya** [comment]
> @aminya I see this now has conflicts that need to be resolved.

I rebased this! The next step is merging #21057

**Comment by sadick254** [comment]
@aminya https://github.com/atom/atom/pull/21057 has been merged. You might need to rebase again.

**Comment by aminya** [comment]
@sadick254 Thanks! I cherry-picked the preparation template. The next step is merging https://github.com/atom/atom/pull/21081!

**Comment by aminya** [comment]
@sadick254 Thanks! I cherry-picked the preparation template. The next step is merging https://github.com/atom/atom/pull/21221

**Comment by aminya** [comment]
@sadick254 This is rebased again. The next one is https://github.com/atom/atom/pull/21226

**Comment by aminya** [comment]
@sadick254 

This is rebased again. The next one is the build template. 
https://github.com/atom/atom/pull/21230

**Comment by aminya** [comment]
@sadick254 

This is rebased again. The next one is the build template. 
https://github.com/atom/atom/pull/21230

**Comment by aminya** [comment]
@sadick254 The next one is #21262

**Comment by aminya** [comment]
@sadick254 This is rebased. I think this can go in one PR. It has the test-template + miscellaneous changes (in the last 4 commits).

Let me know if you want me to break this into two PR.

**Comment by sadick254** [comment]
@aminya Breaking it to two PR's would be great.

**Comment by sadick254** [comment]
This needs another rebasing. @aminya

**Comment by aminya** [comment]
@sadick254 Done! Now, this can be merged.

File: script/vsts/platforms/templates/get-release-version.yml [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+
+jobs:
+
+- job: GetReleaseVersion
+  pool:
+    vmImage: 'ubuntu-latest'
+  steps:
+    - script: |
+        cd script/vsts
+        npm ci
```
Comment (aminya) [inline script/vsts/platforms/templates/get-release-version.yml:17]: I noticed that this is `npm ci` which reverts your change to npm install change. @sadick254.

**Comment by aminya** [comment]
@sadick254 Done! Now, this can be merged.

File: script/vsts/platforms/templates/get-release-version.yml (hunk 1 above)
Comment (DeeDeeG) [reply script/vsts/platforms/templates/get-release-version.yml]: Very briefly: I'd advocate for using `npm ci`.

---

@sadick254 If I recall correctly, the `npm ci` --> `npm install` change was meant to fix intermittent Linux build failures.

It turned out `npm install` doesn't fix those Linux failures. Node 12.16.3 in CI seems to be doing okay for fixing the Linux build failures. (So there isn't a true purpose left as for why to prefer `npm install` over `npm ci`, as I understand it.)

In my personal view, it is slightly advantageous to use `npm ci` iwherever we can. Just a bit faster and cleaner than `npm install`. I admit, though, the speedup is only a few seconds most of the time.

Comment (DeeDeeG) [reply script/vsts/platforms/templates/get-release-version.yml]: But don't worry about that if it slows down merging this. The stakes are so low as to not really matter. To be quite honest.

**Review by sadick254 — APPROVED** [review]
This looks great. Thanks @aminya and @DeeDeeG for the great collaboration on this one. 👍

**Pull request merged**